staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -3645,7 +3645,18 @@ type NodeSpec struct {
 // NodeConfigSource specifies a source of node configuration. Exactly one subfield (excluding metadata) must be non-nil.
 type NodeConfigSource struct {
 	metav1.TypeMeta `json:",inline"`
-	ConfigMapRef    *ObjectReference `json:"configMapRef,omitempty" protobuf:"bytes,1,opt,name=configMapRef"`
+
+	// +k8s:deprecated=configMapRef,protobuf=1
+
+	// ConfigMap is a reference to a Node's ConfigMap
+	ConfigMap *ConfigMapNodeConfigSource `json:"configMap,omitempty" protobuf:"bytes,2,opt,name=configMap"`
+}
+
+type ConfigMapNodeConfigSource struct {
+	// ObjectReference points to the ConfigMap. Name, Namespace, and UID must be specified.
+	ObjectReference `json:",inline" protobuf:"bytes,2,opt,name=objectReference"`
```

Comment:
per https://github.com/kubernetes/kubernetes/issues/61643#issuecomment-380570310, I expected just name and namespace fields, not an inlined ObjectReference

Reply:
I'm prepping another PR that handles the changes discussed in that issue, should I roll that into this one?

Reply:
Even with those changes, in my WIP the Kubelet records UID/ResourceVersion to fully resolve a checkpoint (I'd like to name each checkpoint after a UID/ResourceVersion pair, rather than updating checkpoints in place). In my WIP, this is accomplished by serializing NodeConfigSource to disk, with the ObjectReference also containing UID and ResourceVersion.

Reply:
Basically I'd like everything under the hood to still be fully explicit, I want last-known-good pinned to the specific UID/ResourceVersion that ran well enough to become last-known-good, and I ultimately want the status to report the precise UID/ResourceVersion in use.

Reply:
But in terms of ObjectReference having more fields than necessary, I can probably agree.

Reply:
same comment about exposing fields in the REST API that aren't allowed to be set or aren't honored if they are set. if this type was used in status with uid/resourceVersion filled in, that would be more justifiable (and validation could  prevent setting those two fields in spec, which is still slightly odd, but there's at least a coherent purpose for the fields in the REST API)

Reply:
>if this type was used in status with uid/resourceVersion filled in, that would be more justifiable (and validation could prevent setting those two fields in spec, which is still slightly odd, but there's at least a coherent purpose for the fields in the REST API)

I am in favor of a structured status (mostly because shoving things into only-human-readable node condition strings is much less useful than a structured, machine-readable status), see #56896. I would like the status to be fully explicit. Agree that we should validate they remain unset in spec.

How about I reduce it to name/namespace/uid/resourceVersion fields in this PR, and follow-up with a PR to implement a structured status a-la #56896?

Reply:
sounds good. were you planning to update API validation in this PR? (would prefer it go in with the field changes)

Reply:
Updated the validation; I'll restrict UID in #63221 with the associated changes to resolve #61643.

Reply:
Structured status PR: https://github.com/kubernetes/kubernetes/pull/63314